[prototype] zcfgreconciler: setup non-cancellable, singleton reconciliation job

arulajmani opened the pull request:
See individual commits for details.

arulajmani committed:
jobs: add the zcfg reconciliation job as a new job type

Release note: None

arulajmani committed:
zcfgreconciler: setup non-cancellable, singleton reconciliation job

This patch introduces the `ZcfgReconciliationManager` which is
responsible for ensuring one and only one zone config reconciliation
job exists. Every SQL pod creates one of these on startup. On startup,
the manager checks to see if a `ZcfgReconciliationJob` exists.
If it doesn't, it starts one up and marks it non-cancellable.

This patch creates and assigns a synthetic jobID to the
ZcfgReconciliationJob to ensure singleton semantics. The job resumer,
for now, is a no-op.

References #66680

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/66806)
<!-- Reviewable:end -->

irfansharif committed:
jobs,zcfgreconciler: avoid static zone config job ID

We can check to see that no other jobs with the same type exist.
Additionally denote the job as an automatic one, visible only through
`SHOW AUTOMATIC JOBS`.

Release note: None

irfansharif reviewed (approved):
Appended a commit to unabbreviate "zcfg" in code, and to avoid the static job ID.

arulajmani committed:
jobs: fix test fallout from introducing auto zone config job

Release note: None
